packages/next/src/server/testing/config-testing-utils.ts

```diff
@@ -0,0 +1,155 @@
+import type { IncomingHttpHeaders } from 'node:http'
+import { parse, type UrlWithParsedQuery } from 'node:url'
+import { match } from 'next/dist/compiled/path-to-regexp'
+import {
+  matchHas,
+  prepareDestination,
+} from '../../shared/lib/router/utils/prepare-destination'
+import { buildCustomRoute } from '../../lib/build-custom-route'
+import loadCustomRoutes from '../../lib/load-custom-routes'
+import type { NextConfig } from '../config-shared'
+import { NextResponse } from '../web/exports'
+import { getRedirectStatus } from '../../lib/redirect-status'
+import type {
+  ManifestHeaderRoute,
+  ManifestRedirectRoute,
+  ManifestRewriteRoute,
+} from '../../build'
+import type { BaseNextRequest } from '../base-http'
+import { constructRequest } from './utils'
+import type { Params } from '../request/params'
+
+/**
+ * Tries to match the current request against the provided route. If there is
+ * a match, it returns the params extracted from the path. If not, it returns
+ * undefined.
+ */
+function matchRoute(
+  route: ManifestHeaderRoute | ManifestRedirectRoute | ManifestRewriteRoute,
+  request: BaseNextRequest,
+  parsedUrl: UrlWithParsedQuery
```

Comment:
maybe this should accept list of `filesystemPaths` as those take priority depending on the type of route being matched

Reply:
I like that, but would it be too much work for the user to have to specify the filesystem paths manually? It also looks like the routing code that uses the filesystem is also intertwined with the actual file system so it feels like it might be more difficult to get the same routing behavior? Let me know since you're more familiar with the code.

Reply:
We could potentially have handling to grab these outputs ourselves as long as they provide the project base dir and if no dir provided they can manually provide a list which could default to empty. Can be looked at in a follow-up though